pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1053 @@
+use sp_runtime::{
+	traits::{EnsureAdd, EnsureAddAssign, Zero},
+	DispatchError, DispatchResult,
+};
+
+use crate::{
+	errors::{InvestError, RedeemError},
+	types::{
+		InnerRedeemState, InvestState, InvestTransition, RedeemState, RedeemTransition,
+		TokenSwapReason,
+	},
+	CollectedRedemptionTrancheTokens, Config, Error, Event, ForeignInvestmentInfo,
+	ForeignInvestmentInfoOf, InvestmentState, Pallet, RedemptionState, SwapOf, TokenSwapOrderIds,
+};
+
+mod invest;
+mod redeem;
+
+// Hook execution for (partially) fulfilled token swaps which should be consumed
+// by `TokenSwaps`.
+impl<T: Config> StatusNotificationHook for Pallet<T> {
+	type Error = DispatchError;
+	type Id = T::TokenSwapOrderId;
+	type Status = SwapOf<T>;
+
+	fn notify_status_change(
+		id: T::TokenSwapOrderId,
+		status: SwapOf<T>,
+	) -> Result<(), DispatchError> {
+		let info = ForeignInvestmentInfo::<T>::get(id).ok_or(Error::<T>::InvestmentInfoNotFound)?;
+		let reason = info
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		match reason {
+			TokenSwapReason::Investment => {
+				let pre_state = InvestmentState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(InvestTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(InvestError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_invest_state_transition(&info.owner, info.id, post_state)
+			}
+			TokenSwapReason::Redemption => {
+				let pre_state = RedemptionState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(RedeemTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(RedeemError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_redeem_state_transition(&info.owner, info.id, post_state)
+			}
+		}
+	}
+}
+
+impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
+	type Amount = T::Balance;
+	type CollectInvestResult = ExecutedForeignCollectInvest<T::Balance>;
+	type CurrencyId = T::CurrencyId;
+	type Error = DispatchError;
+	type InvestmentId = T::InvestmentId;
+
+	#[transactional]
+	fn increase_foreign_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		amount: T::Balance,
+		foreign_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		let pre_state = InvestmentState::<T>::get(who, investment_id).unwrap_or_default();
+
+		// NOTE: This adds one db read but let's be safe for the MVP, can hopefully be
+		// removed before deploying to production
+		let unprocessed_invest_amount = T::Investment::investment(who, investment_id)?;
```

Comment:
Same problem here as with the one in `nuge_*`. 

Given this is called this way, I assume the API is misleading. We should maybe think about automatically collecting when calling this. But I am not 100% sure what is best behaviours. WDYT? @lemunozm @NunoAlexandre @wischli

Reply:
Given my comment below we should probably assume, the user must always collect manually? Still not sure TH.

Reply:
I actually had to remove the `nudges` entirely, as their effect occurs upon collecting which we can assume will rather be done by our bot for the time being.